### PR TITLE
Changed return type of randint64_cpu to int64_t to prevent codegen is…

### LINKED ISSUE #117435: Failure to compile a function that takes a remainder of randint result
## Issue description

As in the title.

## Code example

```python
>>> import torch
>>> @torch.compile
... def fn(n):
...     return torch.randint(10, (n,), dtype=torch.int64) % 5
... 
>>> fn(10)
```

<details>

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/eval_frame.py", line 417, in _fn
    return fn(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/eval_frame.py", line 580, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 741, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 384, in _convert_frame_assert
    return _compile(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 643, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 524, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
    transformations(instructions, code_options)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 151, in _fn
    return fn(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/convert_frame.py", line 489, in transform
    tracer.run()
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/symbolic_convert.py", line 2098, in run
    super().run()
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/symbolic_convert.py", line 780, in run
    and self.step()
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/symbolic_convert.py", line 743, in step
    getattr(self, inst.opname)(inst)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/symbolic_convert.py", line 2216, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/output_graph.py", line 914, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/output_graph.py", line 1085, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/output_graph.py", line 1157, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/output_graph.py", line 1138, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/__init__.py", line 1714, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/compile_fx.py", line 1177, in compile_fx
    return aot_autograd(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_functorch/aot_autograd.py", line 889, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_functorch/aot_autograd.py", line 602, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 434, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 639, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 97, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/compile_fx.py", line 1105, in fw_compiler_base
    return inner_compile(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/repro/after_aot.py", line 83, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/debug.py", line 304, in inner
    return fn(*args, **kwargs)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/compile_fx.py", line 317, in compile_fx_inner
    compiled_graph = fx_codegen_and_compile(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/compile_fx.py", line 551, in fx_codegen_and_compile
    compiled_fn = graph.compile_to_fn()
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/graph.py", line 1153, in compile_to_fn
    return self.compile_to_module().call
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_dynamo/utils.py", line 247, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/graph.py", line 1105, in compile_to_module
    mod = PyCodeCache.load_by_key_path(
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/codecache.py", line 1874, in load_by_key_path
    exec(code, mod.__dict__, mod.__dict__)
  File "/tmp/torchinductor_pearu/z3/cz3wydts3yfstrmc67yotclav5b57l4aqpjjn5mwlyaxlfjx3xuq.py", line 47, in <module>
    async_compile.wait(globals())
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/codecache.py", line 2454, in wait
    scope[key] = result.result()
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/concurrent/futures/_base.py", line 440, in result
    return self.__get_result()
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/codecache.py", line 2425, in task
    return CppCodeCache.load(source_code).kernel
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/codecache.py", line 1826, in load
    compile_file(input_path, output_path, cmd)
  File "/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/_inductor/codecache.py", line 1779, in compile_file
    raise exc.CppCompileError(cmd, output) from e
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
CppCompileError: C++ compile error

Command:
/home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/x86_64-conda-linux-gnu-c++ /tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.cpp -shared -fPIC -Wall -std=c++17 -Wno-unused-variable -Wno-unknown-pragmas -D_GLIBCXX_USE_CXX11_ABI=1 -I/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/include -I/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/include/torch/csrc/api/include -I/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/include/TH -I/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/include/THC -I/home/pearu/miniconda3/envs/pytorch-cuda-dev/include/python3.9 -L/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/lib -L/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib -L/home/pearu/git/pytorch/pytorch-sparse-compressed/torch/lib -ltorch -ltorch_cpu -lgomp -ltorch_python -lc10 -mavx512f -mavx512dq -mavx512vl -mavx512bw -mfma -DCPU_CAPABILITY_AVX512 -O3 -DNDEBUG -ffast-math -fno-finite-math-only -fno-unsafe-math-optimizations -march=native -fopenmp -D C10_USING_CUSTOM_GENERATED_MACROS -o /tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.so
```

</details>

```
Output:
In file included from /tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.cpp:2:
/tmp/torchinductor_pearu/2s/c2sj577gdsgvjjryokkxxttxnonx35diffsknzvdsmcs6deenmwx.h: In function 'at::vec::CPU_CAPABILITY::Vectorized<float> to_float_mask(int)':
/tmp/torchinductor_pearu/2s/c2sj577gdsgvjjryokkxxttxnonx35diffsknzvdsmcs6deenmwx.h:408:4: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  408 |   *(uint32_t*)&mask = src ? 0xFFFFFFFF : 0;
      |    ^~~~~~~~~~~~~~~~
/tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.cpp: In function 'void kernel(const long int*, long int*)':
/tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.cpp:13:28: error: no matching function for call to 'mod(long unsigned int&, long int&)'
   13 |             auto tmp4 = mod(tmp2, tmp3);
      |                         ~~~^~~~~~~~~~~~
/tmp/torchinductor_pearu/2s/c2sj577gdsgvjjryokkxxttxnonx35diffsknzvdsmcs6deenmwx.h:147:32: note: candidate: 'template<class T> T mod(T, T)'
  147 | template <typename T> inline T mod(T a, T b) { return a % b; }
      |                                ^~~
/tmp/torchinductor_pearu/2s/c2sj577gdsgvjjryokkxxttxnonx35diffsknzvdsmcs6deenmwx.h:147:32: note:   template argument deduction/substitution failed:
/tmp/torchinductor_pearu/q7/cq7v3kvgvj3lfliputnqep3zoxr7yjde6njkazetae47fgrknbkw.cpp:13:28: note:   deduced conflicting types for parameter 'T' ('long unsigned int' and 'long int')
   13 |             auto tmp4 = mod(tmp2, tmp3);
      |                         ~~~^~~~~~~~~~~~
<snip>
```

## Discussion

While the following patch fixes the issue
```
diff --git a/torch/_inductor/codegen/cpp_prefix.h b/torch/_inductor/codegen/cpp_prefix.h
index 98e3f58245d..76562e749ca 100644
--- a/torch/_inductor/codegen/cpp_prefix.h
+++ b/torch/_inductor/codegen/cpp_prefix.h
@@ -147,6 +147,7 @@ Welford<scalar_t> welford_vec_reduce_all(Welford<at::vec::Vectorized<scalar_t>>
 template <typename T> inline T mod(T a, T b) { return a % b; }
 template <> inline float mod(float a, float b) { return std::fmod(a, b); }
 template <> inline double mod(double a, double b) { return std::fmod(a, b); }
+template <typename T, typename U> inline U mod(T a, U b) { return a % b; }
 
 template <typename scalar_t>
 inline scalar_t max_propagate_nan(scalar_t a, scalar_t b) {
```
it does not answer the question of why the usage of `randint` generates code containing variables using `uint64_t` type. There may be a deeper problem that needs to be fixed. Therefore, applying the above patch is not recommended (nor likely needed) until the cause of the real problem is discovered.

## System Info

- PyTorch version: main

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### PR DESCRIPTION
**Pull request by tringwald**
…sues.

Fixes #117435.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Test failures

File: torch/_inductor/codegen/cpp_prefix.h [hunk 1]
```diff
@@ -179,12 +179,12 @@ float randn_cpu(uint32_t seed, uint32_t offset) {
...
 }
 
-uint64_t randint64_cpu(uint32_t seed, uint32_t offset, int64_t low, int64_t high) {
+int64_t randint64_cpu(uint32_t seed, uint32_t offset, int64_t low, int64_t high) {
   auto gen = at::Philox4_32(seed, 0, offset);
   uint64_t r0 = gen();
   uint64_t r1 = gen();
   uint64_t result = r0 | (r1 << 32);
-  return (result % static_cast<uint64_t>(high - low)) + low;
+  return static_cast<int64_t>(result % (high - low)) + low;
```
Comment (ezyang) [inline torch/_inductor/codegen/cpp_prefix.h:187]: The new version of the code looks suspicious. Due to promotion rules when mixing unsigned and signed quantities, `high - low` gets promoted to `uint64_t`. Fortunately high - low is guaranteed to be non-negative, so this never overflows. So I guess this is OK.

**Review by ezyang — APPROVED** [review]

**Comment by tringwald** [comment]
Rebased on main to fix the CI failures.

**Comment by tringwald** [comment]
Not really sure why that one MacOS build failed. Seems to be a compile error, but the error message is not really helpful.

**Comment by ezyang** [comment]
Maybe the error is

```
 dyld[85187]: Assertion failed: (this->magic == kMagic), function matchesPath, file Loader.cpp, line 154.
```

???

**Comment by tringwald** [comment]
I rebased on main without any further changes and now the MacOS build is green too. The two other failures are unrelated.

File: torch/_inductor/codegen/cpp_prefix.h [hunk 2]
```diff
@@ -144,7 +144,7 @@ Welford<scalar_t> welford_vec_reduce_all(Welford<at::vec::Vectorized<scalar_t>>
 #endif
 
 
-template <typename T> inline T mod(T a, T b) { return a % b; }
+template <typename T, typename U> inline typename std::common_type<T, U>::type mod(T a, U b) { return a % b; }
```
Comment (pearu) [inline torch/_inductor/codegen/cpp_prefix.h:147]: This change was unnecessary for the issue fix.

Comment (tringwald) [reply torch/_inductor/codegen/cpp_prefix.h:147]: IIRC MacOS builds didn't work without this change.

Comment (pearu) [reply torch/_inductor/codegen/cpp_prefix.h:147]: IIUC, this change should not affect builds as the file is used only in runtime.

Comment (tringwald) [reply torch/_inductor/codegen/cpp_prefix.h:147]: With build I meant the full CI run, which includes tests referencing that template.

Comment (pearu) [reply torch/_inductor/codegen/cpp_prefix.h:147]: Is it possible to still see the related CI errors? While I don't mind the change itself, I'd like to understand better why this change was needed: e.g. why wasn't this needed before? And, if this need was introduces by the new test case, perhaps something is still broken in the codegen side?

Comment (tringwald) [reply torch/_inductor/codegen/cpp_prefix.h:147]: I don't think the logs are available anymore, but you could reproduce them by using the 2nd commit of this PR. It was a similar error message as in your initial issue report. clang complained about conflicting types (long long vs long, I think). 
As for why this wasn't needed: I assume this combination was never tested, i.e. clang + mod(randint with int64_t, Python scalar).

Comment (tringwald) [reply torch/_inductor/codegen/cpp_prefix.h:147]: I put up a dummy PR to see if the issue persists: https://github.com/pytorch/pytorch/pull/117842

Comment (tringwald) [reply torch/_inductor/codegen/cpp_prefix.h:147]: > I put up a dummy PR to see if the issue persists: #117842

Seems like the MacOS runs are never triggered that way.

**Pull request closed without merging**